netty/src/main/java/io/grpc/netty/GrpcHttp2HeadersDecoder.java

```diff
@@ -0,0 +1,499 @@
+        int offset1, int length1) {
+      if (length0 != length1) {
+        return false;
+      }
+      return PlatformDependent.equals(bytes0, offset0, bytes1, offset1, length0);
+    }
+
+    protected static byte[] bytes(AsciiString str) {
+      return str.isEntireArrayUsed() ? str.array() : str.toByteArray();
+    }
+
+    protected static AsciiString requireAsciiString(CharSequence cs) {
+      if (!(cs instanceof AsciiString)) {
+        throw new IllegalArgumentException("AsciiString expected. Was: " + cs.getClass().getName());
+      }
+      return (AsciiString) cs;
+    }
+
+    protected static boolean isPseudoHeader(AsciiString str) {
+      return !str.isEmpty() && str.charAt(0) == ':';
+    }
+
+    protected AsciiString validateName(AsciiString str) {
+      int offset = str.arrayOffset();
+      int length = str.length();
+      final byte[] data = str.array();
+      for (int i = offset; i < offset + length; i++) {
+        if (isUpperCase(data[i])) {
+          PlatformDependent.throwException(connectionError(PROTOCOL_ERROR,
+              "invalid header name '%s'", str));
+        }
+      }
+      return str;
+    }
+
+    private void expandHeadersAndValues() {
+      int newValuesLen = values.length + (values.length >> 1);
+      int newNamesAndValuesLen = newValuesLen << 1;
+
+      byte[][] newNamesAndValues = new byte[newNamesAndValuesLen][];
+      AsciiString[] newValues = new AsciiString[newValuesLen];
+      System.arraycopy(namesAndValues, 0, newNamesAndValues, 0, namesAndValues.length);
+      System.arraycopy(values, 0, newValues, 0, values.length);
+      namesAndValues = newNamesAndValues;
+      values = newValues;
+    }
+
+    @Override
+    public int size() {
+      return numHeaders();
+    }
+  }
+
+  /**
+   * A {@link GrpcHttp2InboundHeaders} implementation, optimized for HTTP/2 request headers. That
+   * is, HTTP/2 request pseudo headers are stored in dedicated fields and are NOT part of the
+   * array returned by {@link #namesAndValues()}.
+   *
+   * <p>This class only implements the methods used by {@link NettyServerHandler} and tests. All
+   * other methods throw an {@link UnsupportedOperationException}.
+   */
+  static final class GrpcHttp2RequestHeaders extends GrpcHttp2InboundHeaders {
+
+    private static final AsciiString PATH_HEADER = AsciiString.of(":path");
+    private static final AsciiString AUTHORITY_HEADER = AsciiString.of(":authority");
+    private static final AsciiString METHOD_HEADER = AsciiString.of(":method");
+    private static final AsciiString SCHEME_HEADER = AsciiString.of(":scheme");
+
+    private AsciiString path;
+    private AsciiString authority;
+    private AsciiString method;
+    private AsciiString scheme;
+    private AsciiString te;
+
+    GrpcHttp2RequestHeaders(int numHeadersGuess) {
+      super(numHeadersGuess);
+    }
+
+    @Override
+    public Http2Headers add(CharSequence csName, CharSequence csValue) {
```

Comment:
That's called in in the `Decoder`. I checked, all methods in this class get inlined.